### PR TITLE
MINOR: update JavaDoc for DSL PAPI-API

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
@guozhangwang @dguy @enothereska

**Comment by mjsax** [comment]
Review request for 0.10.2 @guozhangwang @dguy @enothereska

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -117,7 +118,7 @@
      * });
      * }</pre>
      * <p>
-     * The provided {@link KeyValueMapper} must return a {@link KeyValue} type and the return value must not be {@code null}.
+     * The provided {@link KeyValueMapper} must return a {@link KeyValue} type and must not return {@code null}.
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:121]: Hmm good point. Maybe we should consider ignore nulls in maps if users found it not very friendly. Anyways, just a reminder for future changes.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java [hunk 2]
```diff
@@ -17,46 +17,96 @@
...
+     * @return must return {@code null}&mdash;otherwise, and {@link org.apache.kafka.streams.errors.StreamsException
+     * exception} will be thrown
      */
-    R punctuate(final long timestamp);
+    VR punctuate(final long timestamp);
 
     /**
      * Close this processor and clean up any resources.
+     * <p>
+     * It is not possible to return any new output records within {@code punctuate}.
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: Should this line be moved above `punctuate`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java:121]: I would not change anything with regard to `null`: if users want to change "cardinality" they should use "flatMap". If we want to do some reasoning about performance, it's good to know that map does not change cardinality: one in, one out.

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: No. Must be `It is not possible to return any new output records within {@code close}.`

**Comment by mjsax** [comment]
@guozhangwang Updated.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 3]
```diff
@@ -17,48 +17,93 @@
 
 package org.apache.kafka.streams.kstream;
 
+import org.apache.kafka.common.annotation.InterfaceStability;
 import org.apache.kafka.streams.processor.ProcessorContext;
 
 /**
- * A stateful {@link Transformer} interface to transform a key-value pair into a new value.
+ * The {@link Transformer} interface for stateful mapping of an input record to a zero, one, or multiple new output
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: remove 'a'

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 4]
```diff
@@ -17,48 +17,93 @@
...
+ * record (both key and value type can be altered arbitrarily).
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: records

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 5]
```diff
@@ -17,48 +17,93 @@
...
-     * @return new value; if null no key-value pair will be forwarded to down stream
+     * @return new {@link org.apache.kafka.streams.KeyValue key-value pair}&mdash;if {@code null} no key-value pair will
+     * be forwarded to down stream
      */
     R transform(final K key, final V value);
 
     /**
-     * Perform any periodic operations and possibly generate a key, if this processor {@link ProcessorContext#schedule(long) schedules itself} with the context
-     * during {@link #init(ProcessorContext) initialization}.
+     * Perform any periodic operations and possible generate new {@link org.apache.kafka.streams.KeyValue key-value
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: possibly

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 6]
```diff
@@ -17,48 +17,93 @@
...
+     * pairs} if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context during
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: schedules

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 7]
```diff
@@ -17,48 +17,93 @@
...
+     * {@link #init(ProcessorContext) initialization}.
+     * <p>
+     * To generate new {@link org.apache.kafka.streams.KeyValue key-value pairs}
+     * {@link ProcessorContext#forward(Object, Object)}, {@link ProcessorContext#forward(Object, Object, int)}, and
+     * {@link ProcessorContext#forward(Object, Object, String)} can be used.
+     * <p>
+     * Note, that {@code punctuate} is called base on <it>stream time</it> (i.e., time progress with regard to
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: based

Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: progresses

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 8]
```diff
@@ -17,48 +17,93 @@
...
+     * timestamps return by the used {@link org.apache.kafka.streams.processor.TimestampExtractor TimestampExtractor})
+     * and not based on wall-clock time.
      *
-     * @param timestamp the stream time when this method is being called
-     * @return new value; if null it will not be forwarded to down stream
+     * @param timestamp the stream time when {@code punctuate} is being called
+     * @return must return {@code null}&mdash;otherwise, and {@link org.apache.kafka.streams.errors.StreamsException
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: and -> a

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java [hunk 9]
```diff
@@ -17,46 +17,96 @@
...
+     * Perform any periodic operations if this processor {@link ProcessorContext#schedule(long) schedule itself} with
+     * the context during {@link #init(ProcessorContext) initialization}.
+     * <p>
+     * It is not possible to return any new output records within {@code punctuate}.
+     * Using {@link ProcessorContext#forward(Object, Object)}, {@link ProcessorContext#forward(Object, Object, int)},
+     * or {@link ProcessorContext#forward(Object, Object, String)} will result in an
+     * {@link org.apache.kafka.streams.errors.StreamsException exception}.
+     * Furthermore, {@code punctuate} must return {@code null}.
+     * <p>
+     * Note, that {@code punctuate} is called base on <it>stream time</it> (i.e., time progress with regard to
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: base -> based
progress -> progressed

File: streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java [hunk 10]
```diff
@@ -17,46 +17,96 @@
...
+     * timestamps return by the used {@link org.apache.kafka.streams.processor.TimestampExtractor TimestampExtractor})
+     * and not based on wall-clock time.
      *
-     * @param timestamp the stream time when this method is being called
-     * @return new value; if null it will not be forwarded to down stream
+     * @param timestamp the stream time when {@code punctuate} is being called
+     * @return must return {@code null}&mdash;otherwise, and {@link org.apache.kafka.streams.errors.StreamsException
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/ValueTransformer.java]: and -> a

**Review by dguy — CHANGES_REQUESTED** [review]
Just some typos

**Comment by mjsax** [comment]
Updated.

**Comment by guozhangwang** [comment]
Merged to trunk and 0.10.2.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 11]
```diff
@@ -17,48 +17,96 @@
...
+     * {@link ProcessorContext#forward(Object, Object, String)} can be used.
+     * <p>
+     * Note that {@code punctuate} is called based on <it>stream time</it> (i.e., time progresses with regard to
+     * timestamps return by the used {@link TimestampExtractor})
+     * and not based on wall-clock time.
      *
-     * @param timestamp the stream time when this method is being called
-     * @return new value; if null it will not be forwarded to down stream
+     * @param timestamp the stream time when {@code punctuate} is being called
+     * @return must return {@code null}&mdash;otherwise, a {@link StreamsException exception} will be thrown
```
Comment (mihbor) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java:100]: I don't think this change is correct. KStreamTransformProcessor does:
if (pair != null) context().forward(pair.key, pair.value);
so to the previous @return comment was correct

https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamTransform.java#L64-L67

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java:100]: Thanks for reporting this @mihbor (I guess this was a mix up with `ValueTransformer`). Can you open an PR to fix it?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java:100]: Thanks for the PR!

**Pull request closed without merging**